cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -555,6 +562,57 @@ Future<std::shared_ptr<parquet::arrow::FileReader>> ParquetFileFormat::GetReader
       });
 }
 
+struct CastingGenerator {
```

Comment:
I'm a bit curious why casting generator is required, since during reading parquet to arrow, Parquet reader already applies a round of casting.

Reply:
You're probably right and it likely does perform some casts during the read. However, it doesn't seem to be doing it for certain cases like String to LargeString.
Could you point me to where the Parquet Reader should be performing this cast? I'm happy to add it there.

Reply:
https://github.com/apache/arrow/blob/main/cpp/src/parquet/arrow/reader_internal.cc#L474

Refer to here

Reply:
Thank you!
Based on what I see, that is only responsible for casting the data to the logical type specified in the parquet metadata and not the Arrow type we want to convert to (the one in the dataset_schema). For strings, that seems to always map to a String type (based on `FromByteArray` which is called by `GetArrowType` which is called by `GetTypeForNode` which is called by `NodeToSchemaField` which is called in `SchemaManifest::Make` during the creation of the `LeafReader`). Am I missing something?

Reply:
> Based on what I see, that is only responsible for casting the data to the logical type specified in the parquet metadata and not the Arrow type we want to convert to (the one in the dataset_schema)

Parquet logical type doesn't have an arrow schema, isn't it? Binary reader reads from `::arrow::BinaryBuilder`, and casting it to user-specified binary type.

> For strings, that seems to always map to a String type (based on FromByteArray which is called by GetArrowType which is called by GetTypeForNode which is called by NodeToSchemaField which is called in SchemaManifest::Make during the creation of the LeafReader).

Yeah, you're right, the read "cast" with file-schema rather than an expected schema. I think a native cast is better here but this doesn't solve your problem, perhaps I can trying to add a naive `SchemaManifest` with hint solving here, but it would spend some time.

Reply:
Provide hint would looks like: https://github.com/apache/arrow-rs/pull/5939

Maybe I can add separate issue for that

Reply:
> Parquet logical type doesn't have an arrow schema, isn't it? 

As far as I understand, the parquet metadata may or may not have the arrow schema. I believe it depends on the writer. It looks like it tries to get that using `GetOriginSchema` in `SchemaManifest::Make`. However, the schema at write time might not be the same as the schema the reader expects.

> Binary reader reads from ::arrow::BinaryBuilder, and casting it to user-specified binary type.

Sorry, I didn't quite follow. Are you saying that we should use this to do the cast at read time somehow?

> I think a native cast is better here but this doesn't solve your problem, perhaps I can trying to add a naive SchemaManifest with hint solving here, but it would spend some time.
> Maybe we should rethink the GetTypeForNode handling for string/large_string/stringView, or using some handle written type hint here.

That makes sense to me.

> Maybe I can add separate issue for that

That would be great, thanks!

Reply:
I just check this. We can first add a `CastingGenerator`, because sometimes we would have schema evolution here, and need cast to a "final type". The Parquet reader code can be regard as an optimization

Reply:
Sounds good, thanks!